src/Workspaces/Core/Desktop/Utilities/RemoteTask.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal interface IRemoteTask<T>
+    {
+        void Cancel();
+        void SetCompletion(IRemoteTaskCompletion<T> completion);
+        T GetResult();
+    }
+
+    internal interface IRemoteTaskCompletion<T>
+    {
+        void Complete(T result);
+        void Fail(Exception exception);
+    }
+
+    internal static class RemoteTask
+    {
+        public static Task<TResult> ToTask<TResult>(this IRemoteTask<TResult> asyncTask, CancellationToken cancellationToken)
+        {
+            return new RemoteTaskCompletion<TResult>(asyncTask, cancellationToken).Task;
+        }
+    }
+
+    internal class RemoteTaskCompletion<T> : MarshalByRefObject, IRemoteTaskCompletion<T>
```

Comment:
Please do some heavy documenting on this and the RemoteTask type of which lives in which AppDomain and how they get created. I've had to debug similar things in the past and it's really hard looking at code to get a high-level sense of 'what should live where'.